javascript/ql/src/Security/CWE-020/IncompleteHostnameRegExp.ql

```diff
@@ -13,16 +13,41 @@
 import javascript
 
 /**
- * A taint tracking configuration for incomplete hostname regular expressions sources.
+ * Gets a node whose value may flow (inter-procedurally) to a position where it is interpreted
+ * as a regular expression.
  */
-class Configuration extends TaintTracking::Configuration {
-  Configuration() { this = "IncompleteHostnameRegExpTracking" }
+DataFlow::Node regExpSource(DataFlow::Node re, DataFlow::TypeBackTracker t) {
+  t.start() and
+  re = result and
+  isInterpretedAsRegExp(result)
+  or
+  exists(DataFlow::TypeBackTracker t2, DataFlow::Node succ | succ = regExpSource(re, t2) |
+    exists(DataFlow::StepSummary summary |
+      DataFlow::StepSummary::smallstep(result, succ, summary) and
+      t = t2.prepend(summary)
+      or
+      result = succ.getAPredecessor() and
+      t = t2
```

Comment:
Would it perhaps be worth to encapsulate these two disjuncts into a predicate on `SourceNode` to sit alongside `backtrack`? Finding a good name needs a bit of thought, but as it stands we are now directly using an internal predicate (`smallstep`) in one of our queries, which is arguably not a precedent we want to set.

Reply:
I have pushed two commits as a proposal. I think they make the API nicer.

I propose to put the predicates on the type tracker classes, and call them `smallstep`. Similarly for `step`: make the `SourceNode::backtrack` predicate a simple forwarder predicate.
See new commit 1/2.
Similar adjustments should be done for the forwards tracking predicate if you agree.

Additionally, the use of the `StepSummary::smallstep` predicate seems nicer if the result is a `StepSummary`. See new commit 2/2.
A similar adjustments should be done for the `StepSummary::step` predicate, but that should be done in another PR, since that requires other refactorings.

These changes require another sanity check of the performance.

Reply:
I'm supportive of not adding more member predicates to `SourceNode`. I like this proposal as a stop-gap solution.

I'll note this could be fixed at the root if `SourceNode` was somehow restored to its original form: nodes without a local predecessor, in particular, including literals and concatenation expressions. Maybe it's worth taking another shot at that - a solution to that would be so much more valuable in the long term.